### PR TITLE
Dta2 119/Refactor estimation pipeline

### PR DESCRIPTION
**Pull request by dexter2206**
## Description

This refactors resource estimation code by splitting it into two phases:

- transforming
- resource estimation

In particular, whether Clifford+T synthesis is performed is controlled by choosing different transformers.

## Please verify that you have completed the following steps

- [x] I have self-reviewed my code.
- [ ] I have included test cases validating introduced feature/fix.
- [ ] I have updated documentation.

### CODE REVIEW

File: examples/qsp_vlasov.py [hunk 1]
```diff
@@ -73,81 +71,47 @@ def main():
...
-        tolerable_logical_error_rate - qsp_required_precision
-    ) / 3
+    error_budget = {
+        "qsp_required_precision": qsp_required_precision,
+        "tolerable_circuit_error_rate": tolerable_logical_error_rate,
+        "total_error": 1e-2,
+        "synthesis_error_rate": 0.5,
+        "ec_error_rate": 0.5,
+    }
+    specs = {
```
Comment (mstechly) [inline examples/qsp_vlasov.py:81]: I think we're not using these specs anywhere in the end – instead we use transformers. Right?

File: src/benchq/resource_estimation/v2/estimators.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+            synthesis_multiplier=synthesis_multiplier,
+            code_distance=code_distance,
+            logical_error_rate=logical_error_rate,
+            max_graph_degree=max_degree,
+            n_nodes=n_nodes,
+            n_measurement_steps=n_measurement_steps,
+            total_time=wall_time,
+        )
+
+    @singledispatchmethod
```
Comment (mstechly) [inline src/benchq/resource_estimation/v2/estimators.py:185]: Is this still needed?

File: src/benchq/resource_estimation/v2/estimators.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+            else self._ec_error_rate_unsynthesized
+        )
+
+        # Change to code distance
+        code_distance = self._minimize_code_distance(
+            n_nodes, error_budget, ec_error_rate
+        )
+
+        max_degree = max(deg for _, deg in graph.degree())
+        logical_operation_error_rate = (
```
Comment (AthenaCaesura) [inline src/benchq/resource_estimation/v2/estimators.py:137]: nitpick: Why not call `_logical_operation_error_rate`?

File: src/benchq/resource_estimation/v2/estimators.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+                program_graph = combine_subcomponent_graphs(problem)
+                return self._estimate_resource_for_graph(
+                    program_graph, n_nodes, problem.synthesized, error_budget
+                )
+            else:
+                raise NotImplementedError(
+                    "Resource estimation without combining subgraphs is not yet "
+                    "supported."
+                )
+            # use dummy graph
```
Comment (AthenaCaesura) [inline src/benchq/resource_estimation/v2/estimators.py:197]: Can this code be deleted?

File: src/benchq/resource_estimation/v2/master.py
Comment (AthenaCaesura) [inline src/benchq/resource_estimation/v2/master.py:1]: This file could probably use a more descriptive name.

File: src/benchq/resource_estimation/v2/master.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+from .transformers import synthesize_clifford_t
+
+
+def run_resource_estimation_pipeline(
+    program_or_circuit,
+    error_budget,
+    estimator,
+    transformer=synthesize_clifford_t,
+):
+    transformed = transformer(program_or_circuit, error_budget)
```
Comment (AthenaCaesura) [inline src/benchq/resource_estimation/v2/master.py:12]: Not sure what a `transformer` is, but @mstechly says it needs a rework anyways so we won't worry about that right now.

**Review by AthenaCaesura — APPROVED** [review]

File: src/benchq/resource_estimation/v2/estimators.py (hunk 3 above)
Comment (mstechly) [reply src/benchq/resource_estimation/v2/estimators.py:137]: Cause it's just a variable name?

**Pull request merged**